crates/agent-sql/src/controllers.rs

```diff
@@ -0,0 +1,172 @@
+use crate::{CatalogType, Id, TextJson};
+use chrono::prelude::*;
+use serde::Serialize;
+use serde_json::value::RawValue;
+use sqlx::types::Uuid;
+use std::fmt::Debug;
+
+#[derive(Debug)]
+pub struct ControllerJob {
+    pub live_spec_id: Id,
+    pub catalog_name: String,
+    pub last_pub_id: Id,
+    pub live_spec: Option<TextJson<Box<RawValue>>>,
+    pub built_spec: Option<TextJson<Box<RawValue>>>,
+    pub spec_type: CatalogType,
+    pub controller_next_run: Option<DateTime<Utc>>,
+    pub controller_version: i32,
+    pub updated_at: DateTime<Utc>,
+    pub logs_token: Uuid,
+    pub status: TextJson<Box<RawValue>>,
+    pub failures: i32,
+    pub error: Option<String>,
+}
+
+#[tracing::instrument(level = "debug", skip(txn))]
+pub async fn dequeue(
+    txn: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    controller_version: i32,
+) -> sqlx::Result<Option<ControllerJob>> {
+    sqlx::query_as!(
+        ControllerJob,
+        r#"with needs_periodic(live_spec_id) as (
+            select id as live_spec_id
+            from live_specs
+            where
+                -- This condition is required in order to for this query to use the sparse index
+                controller_next_run is not null
+                and controller_next_run <= now()
+            order by controller_next_run asc
+        ),
+        needs_upgrade(live_spec_id) as (
+            select live_spec_id
```

Comment:
As noted in a prior comment -- this is gonna do a full table scan every time.

I'm coming to believe that `controller_version` isn't buying us much and should be removed.
* We can always have migrations tweak controller_next_run if we want to do a mass-run of controllers because of a software release.
* The remaining use case for `controller_version` is understanding how to interpret a status (is it the old or new version), but we have lots of experience doing incremental updates of JSON specifications and we've never before needed a version to do so.

It seems like `controller_version`'s utility can be gotten at through other mechanisms and so it should be dropped. If that's wrong and we do find we need it, it's also easy to add it as a future migration with a default.

Reply:
I think `controller_version` is still needed in order to allow for agent rollouts. When we deploy a new agent version (that has an incremented `controller_version`), there will still potentially be running agent processes having an older `controller_version`. Part of the purpose of `controller_version` is to allow agents to filter controller jobs so that they don't run for rows that have already been upgraded.

TBH I was thinking that a full table scan might not be a problem in the short term. In practice, most full table scans on `live_specs` seemed to take somewhere in the ~45ms range, and I'm assuming that `controller_jobs` will be roughly similar or maybe slightly faster (fewer columns). But I haven't tested the `deque` performance under realistic conditions.

If we want to avoid the full table scans, I can think of a few approaches:
- Add an index on `controller_version`
- execute a single query when the agent starts up, which sets `controller_next_run = now()` for all rows having `controller_version < $agent_controller_version`

From my perspective, I'd _really_ like to avoid needing to do a database migration when updating the controller version. And I don't think it'd really work, given the possibility for multiple agent versions to be running simultaneously during rollouts.

Reply:
Following up after a VC, we decided to keep `controller_version`, but just not use it to dequeue jobs needing "upgraded". This actually allows the initial rollout of controllers for existing `live_specs` to be done incrementally. The `dequeue` function will only pick up rows that have _both_ a `controller_next_run` that's due and a corresponding `controller_jobs` row. So we can remove the initial backfill of the `controller_jobs` table from the controllers database migration, and move it into a separate migration, which is run _after_ the rollout of new agents completes. This backfilling of `controller_jobs` rows can then be done incrementally, which limits the blast radius if anything goes wrong.